src/extensions/surveys.tsx

```diff
@@ -187,6 +193,84 @@ export function generateSurveys(posthog: PostHog) {
     }, 3000)
 }
 
+export function usePopupVisibility(
+    survey: Survey,
+    posthog: PostHog | undefined,
+    delay: number,
+    isPreviewMode: boolean
+) {
+    const [isPopupVisible, setIsPopupVisible] = useState(isPreviewMode || delay === 0)
+    const [isSurveySent, setIsSurveySent] = useState(false)
+
+    useEffect(() => {
+        if (isPreviewMode || !posthog) {
+            return
+        }
+
+        const handleSurveyClosed = () => {
+            setIsPopupVisible(false)
+        }
+
+        const handleSurveySent = () => {
+            if (!survey.appearance?.displayThankYouMessage) {
+                return setIsPopupVisible(false)
+            }
+
+            setIsSurveySent(true)
+
+            if (survey.appearance?.autoDisappear) {
+                setTimeout(() => {
+                    setIsPopupVisible(false)
+                }, 5000)
+            }
+        }
+
+        window.addEventListener('PHSurveyClosed', handleSurveyClosed)
+        window.addEventListener('PHSurveySent', handleSurveySent)
+
+        if (delay > 0) {
+            const timeoutId = setTimeout(() => {
```

Comment:
hard for me to reason via the code, but what happens in the case when:

1. We have a popup with delay 5 seconds (surveyA) that is active & matching
2. We have a popup with delay 0 seconds (surveyB) that is active & matching
3. `surveyA` matches first, and we end up here, call setTimeout. `isPopupVisible` is false.
4. `surveyB` matches as well, and is shown directly, because no other survey is visible.
5. 5 seconds later, `surveyA` is shown on top of `surveyB`. I think this will happen because the existing survey visible check happens at a higher level, and not inside the setTimeout here. I think the same problem magnifies when you have multiple surveys with different timeouts.

Or at least that's what I think will happen, haven't run this, try it out please :) I think this delay check needs to go at a higher level to prevent this. Another thing to figure out is how do we want to handle multiple delayed & non-delayed surveys matching - always show the non-delayed ones first, and only then start the timing for delayed surveys (I think I prefer this but haven't thought too much about it), or 'undefined behaviour', whichever ones comes first in the list is set to show first, so if there's a 30s delay, this one will wait 30 seconds, show up, and then any remaining non-delayed ones show up after.

Reply:
This is a thoughtful comment, so I want to make sure I respond to it.

Basically, what I ended up doing was differentiating between the survey that was actively being focused on (I created an action queue of SurveyA, SurveyB, etc, ordered by the delay associated with that survey), and then visibility of the survey itself (which is handled by this `usePopupVisibility` hook.

Here's what happens now:

- we load all surveys from the posthog SDK and throw them into a list, orderd by survey delay (where the non-delayed surveys all appear first, and then proceeding in sequential order with the longest delays at the end of this list)
- we add the first item in this list to the `surveysInFocus` set, which is a Set that has strictly one member, and it keeps track of the survey about which we are evaluating visibility
- we then use `usePopupVisiblity` to determine how to display that survey on the page.

once that survey has completed its popup lifecycle, we then remove it from the `surveysInFocus` set and proceed to the next one.

This way, handling visibility and handling which surveys to evaluate are separate processes, which prevent surveys from colliding with each other.